bolus.py

```diff
@@ -0,0 +1,156 @@
+
+def bolus(start_time, carbs, timesteps, params):
+    bolus_data = []
+    for value, timestamp in zip(carbs, timesteps):      
+        normal_or_square = random.randint(0, 9) 
+        if normal_or_square == 1 or normal_or_square == 2: #2 in 10 are dual square
+            bolus_data.append(dual_square_bolus(value, timestamp, start_time, params))
+        elif normal_or_square == 3: #1 in 10 is a sqaure bolus
+            bolus_data.append(square_bolus(value, timestamp, start_time, params))
+        else: #8 of 10 are normal boluses 
+            bolus_data.append(normal_bolus(value, timestamp, start_time, params))
+    return bolus_data
+
+def dual_square_bolus(value, timestamp, start_time, params):
+    bolus_entry = {}
+    bolus_entry = common_fields.add_common_fields('bolus', bolus_entry, timestamp, params)
+    bolus_entry["subType"] = "dual/square"
+    carb_ratio = get_carb_ratio(start_time, bolus_entry["deviceTime"], params)
+    insulin = int(value) / carb_ratio
+    bolus_entry["normal"] = tools.round_to(random.uniform(insulin / 3, insulin / 2)) 
+    bolus_entry["extended"] = tools.round_to(insulin - bolus_entry["normal"]) 
+    bolus_entry["duration"] = random.randrange(1800000, 5400000, 300000) #in ms
+    
+    interrupt = random.randint(0,9) #interrupt 1 in 10 boluses
+    if interrupt == 1:
+        bolus_entry = (interrupted_dual_square_bolus(bolus_entry["normal"], 
+                       bolus_entry["extended"], bolus_entry["duration"], timestamp, params))
+    return bolus_entry  
+
+def square_bolus(value, timestamp, start_time, params):
+    bolus_entry = {}
+    bolus_entry = common_fields.add_common_fields('bolus', bolus_entry, timestamp, params)
+    bolus_entry["subType"] = "square"
+    bolus_entry["duration"] = random.randrange(1800000, 5400000, 300000)
+    carb_ratio = get_carb_ratio(start_time, bolus_entry["deviceTime"], params)
+    insulin = tools.round_to(int(value) / carb_ratio)
+    bolus_entry["extended"] = insulin
+    return bolus_entry
+
+def normal_bolus(value, timestamp, start_time, params):
+    bolus_entry = {}
+    bolus_entry = common_fields.add_common_fields('bolus', bolus_entry, timestamp, params)
+    bolus_entry["subType"] = "normal"
+    carb_ratio = get_carb_ratio(start_time, bolus_entry["deviceTime"], params)
+    insulin = tools.round_to(int(value) / carb_ratio)
+    bolus_entry["normal"] = insulin
+
+    interrupt = random.randint(0,9) #interrupt 1 in 10 boluses
+    if interrupt == 1:
+        bolus_entry = interrupted_bolus(insulin, timestamp, params)
+    return bolus_entry
+
+def interrupted_bolus(value, timestamp, params):
+    bolus_entry = {}
+    bolus_entry = common_fields.add_common_fields('bolus', bolus_entry, timestamp, params)
+    bolus_entry["expectedNormal"] = value
+    bolus_entry["normal"] = tools.round_to(value - random.uniform(0, value))
+    return bolus_entry
+
+def interrupted_dual_square_bolus(normal, extended, duration, timestamp, params):
```

Comment:
TODO for me: take a closer look at this once I can play around with it in the interpreter because I have a docstring telling me about the params for each function.